Loki: Retry web socket connection when connection is closed abnormally

ivanahuckova opened the pull request:
**How to reproduce the issue**:

See https://github.com/grafana/grafana/issues/29348. 

**What this PR does / why we need it**:
This PR fixes 2 issues:
- When live tailing throws error and the received data/logs are `null`, we use empty `[]` as our data and this way `rowsToRender.slice(100)` [won't throw error.](https://github.com/grafana/grafana/blob/master/public/app/features/explore/LiveLogs.tsx#L132) Before this, we passed `null` and null.slice() threw error as well. 

- When we receive **[1006](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent) error code** which means that connection was closed abnormally (that is, with no close frame being sent). For this specifically, we retry connection. 
**Which issue(s) this PR fixes**:

**Fixed - going over 1h:** 
![image](https://user-images.githubusercontent.com/30407135/100475253-f53ba100-30e2-11eb-88f6-8f1b7d235c87.png)

**Broken:**
![image](https://user-images.githubusercontent.com/30407135/100470844-5199c300-30d9-11eb-8990-b793d498aed8.png)

![image](https://user-images.githubusercontent.com/30407135/100470874-5f4f4880-30d9-11eb-92fc-7147b8afa000.png)

**Fixes:**
Fixes https://github.com/grafana/grafana/issues/29348

**Special notes for your reviewer**:

ivanahuckova committed:
Add retryWhen to Loki socket

ivanahuckova committed:
Refactor,

hugohaggmark commented on public/app/plugins/datasource/loki/live_streams.ts:
```diff
@@ -42,9 +42,19 @@ export class LiveStreams {
+            // Code 1006 is used to indicate that a connection was closed abnormally.
+            // If connection was closed abnormally, we wish to retry, otherwise throw error.
+            if (error.code === 1006) {
+              console.log('Reconnecting websocket...');
```
nit: maybe remove console output?

ivanahuckova replied:
Removed :)

hugohaggmark commented on public/app/plugins/datasource/loki/live_streams.ts:
```diff
@@ -42,9 +42,19 @@ export class LiveStreams {
-      catchError(err => {
-        return throwError(`error: ${err.reason}`);
-      }),
+      retryWhen((attempts: Observable<any>) =>
+        attempts.pipe(
+          mergeMap(error => {
+            // Code 1006 is used to indicate that a connection was closed abnormally.
+            // If connection was closed abnormally, we wish to retry, otherwise throw error.
+            if (error.code === 1006) {
+              console.log('Reconnecting websocket...');
+              return timer(1000);
+            }
+            return throwError(`error: ${error.reason}`);
+          })
+        )
+      ),
```
Skilled work on the RxJs code. The code now will retry into infinity with 1s delay between attempts? Maybe we should limit the retries to x number of retries?

aocenas replied:
Not sure if I read this right and did not find good explanation in the docs but shouldn't it return `timer(1000).take(1)` or something like that? I mean what we want is to wait for error and then wait 1s and retry and then wait for another error and repeat right? But having a timer that will fire every 1s forever after the first error does not seem correct but maybe there is some internal implementation that prevents that which isn't documented?

feels like this section from the [docs](https://www.learnrxjs.io/learn-rxjs/operators/transformation/mergemap) applies here:
> Be aware that because mergeMap maintains multiple active inner subscriptions at once it's possible to create a memory leak through long-lived inner subscriptions. A basic example would be if you were mapping to an observable with an inner timer, or a stream of dom events. In these cases, if you still wish to utilize mergeMap you may want to take advantage of another operator to manage the completion of the inner subscription, think take or takeUntil. You can also limit the number of active inner subscriptions at a time with the concurrent parameter, seen in example 5.

aocenas replied:
Ah ok now I read that `timer(1000)` won't emit indefinitely if the second arg for period is not specified so disregard my previous comment.

hugohaggmark replied:
I still think we should consider adding a max number of attempts counter, but that is just me being super cautious.

ivanahuckova replied:
I have added emitting of warning message `Websocket connection is being disrupted. We keep reconnecting but consider starting new live tailing again.` after we reconnect 10 times and every reconnection after that. So live tailing will be still working, but it is going to be super obvious that something is wrong. Also we have changed the timer from 1s to 5s.

hugohaggmark commented on public/app/plugins/datasource/loki/live_streams.ts:
```diff
@@ -42,9 +42,19 @@ export class LiveStreams {
-      catchError(err => {
-        return throwError(`error: ${err.reason}`);
-      }),
+      retryWhen((attempts: Observable<any>) =>
```
This is an excellent candidate for testing using `expect(getStream()).toEmitValuesWith()`. Maybe you could pass webSocket as an argument to the function so you can pass your own Observable when testing?.

hugohaggmark reviewed (commented):
Looks great ⭐ , I added some comments that you might want to consider.

davkal commented:
If retries are cheap, we can keep retrying every 5-ish seconds, and keep trying. If it keeps failing, then show a message to the user but keep retrying and tell the user that something is disrupted but that it keeps retrying.

ivanahuckova committed:
Add info message after 10 reconnections, remove console log

ivanahuckova committed:
Add test

ivanahuckova commented:
@aocenas @hugohaggmark @davkal I have added test and implemented functionality as we discussed. Thanks Hugo so much for your help writing the test. 🙏

hugohaggmark commented on public/app/plugins/datasource/loki/live_streams.test.ts:
```diff
@@ -105,6 +106,39 @@ describe('Live Stream Tests', () => {
+    await expect(liveStreams.getStream(makeTarget('url_to_match'), 100)).toEmitValuesWith(received => {
+      const dataBefore = received[0];
+      const dataAfter = received[1];
+      expect(dataBefore).not.toBe(undefined);
+      expect(dataAfter).not.toBe(undefined);
```
In order to test that our retry logic works as expected, maybe we should check for specific log rows instead of just toBeDefined?

ivanahuckova replied:
Yes, makes sense! Done in https://github.com/grafana/grafana/pull/29438/commits/e5968b923eb1c957d26800357991e0d6ea02ef8f

hugohaggmark replied:
Beautiful ⭐

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivana/loki-websockets

ivanahuckova committed:
Specify response in tests

ivanahuckova committed:
Check specific lines for content

ivanahuckova committed:
Remove unused imports

hugohaggmark commented on public/app/plugins/datasource/loki/live_streams.ts:
```diff
@@ -42,9 +43,26 @@ export class LiveStreams {
+            if (error.code === 1006) {
+              if (retryAttempt > 10) {
+                // If more than 10 times retried, show user warning, but keep reconnecting
+                appEvents.emit(AppEvents.alertWarning, [
+                  'Websocket connection is being disrupted. We keep reconnecting but consider starting new live tailing again.',
+                ]);
```
nit: Not sure you're supposed to use appEvents outside of the main app.

torkelo replied:
Yea better to return a data error here and let Grafana show that

hugohaggmark replied:
That would work if we didn't need to retry, in this case we only want to show information to the user but not stop the retries.

ivanahuckova replied:
@hugohaggmark as mentioned on our call I have added console.warn and created an issue for informing user in better way https://github.com/grafana/grafana/issues/29609. As we can't inform the user about issues with their connection, if you think it makes sense to add max retries (e.g. 30?), I can do that.

hugohaggmark replied:
Yeah, I think that is a good idea

hugohaggmark reviewed (approved):
This looks great to me ⭐

ivanahuckova committed:
Replace AppEvents with console.warn

ivanahuckova committed:
Add limit of re-connections
